tests/unit/providers/rootio/test_rootio.py

```diff
@@ -0,0 +1,221 @@
+
+    @pytest.fixture()
+    def workspace_dir(self, tmp_path):
+        ws = workspace.Workspace(root=str(tmp_path / "rootio"), name="rootio")
+        ws.create()
+        return ws
+
+    def test_parser_emit_unaffected_for_os_packages(self, workspace_dir, mock_vulnerability_data):
+        """Test that parser emits ROOTIO_UNAFFECTED markers for OS packages"""
+        p = parser.Parser(workspace=workspace_dir)
+        
+        # Mock the API response
+        with patch.object(parser.Parser, "_fetch_vulnerabilities", return_value=mock_vulnerability_data):
+            p._process_vulnerabilities(mock_vulnerability_data)
+            
+            # Check that results were written
+            results_path = workspace_dir.results_path
+            assert os.path.exists(results_path)
+            
+            # Verify the output contains ROOTIO_UNAFFECTED markers
+            vuln_files = list(results_path.glob("*.json"))
+            assert len(vuln_files) > 0
+            
+            # Check for CVE-2023-1234 (OS package vulnerability)
+            found_unaffected = False
+            for vuln_file in vuln_files:
+                with open(vuln_file) as f:
+                    data = json.load(f)
+                    if data.get("Vulnerability", {}).get("Name") == "CVE-2023-1234":
+                        for fixed_in in data["Vulnerability"].get("FixedIn", []):
+                            if fixed_in.get("Version") == "ROOTIO_UNAFFECTED":
+                                found_unaffected = True
+                                assert fixed_in.get("VulnerableRange") == "NOT version_contains .root.io"
+            
+            assert found_unaffected, "Should have ROOTIO_UNAFFECTED marker for OS packages"
+
+    def test_parser_emit_unaffected_for_language_packages(self, workspace_dir, mock_vulnerability_data):
+        """Test that parser emits ROOTIO_UNAFFECTED markers for language packages"""
+        p = parser.Parser(workspace=workspace_dir)
+        
+        # Mock the API response
+        with patch.object(parser.Parser, "_fetch_vulnerabilities", return_value=mock_vulnerability_data):
+            p._process_vulnerabilities(mock_vulnerability_data)
+            
+            # Check for CVE-2023-5678 (Python package vulnerability)
+            results_path = workspace_dir.results_path
+            vuln_files = list(results_path.glob("*.json"))
+            
+            found_python_unaffected = False
+            for vuln_file in vuln_files:
+                with open(vuln_file) as f:
+                    data = json.load(f)
+                    if data.get("Vulnerability", {}).get("Name") == "CVE-2023-5678":
+                        namespace = data["Vulnerability"].get("NamespaceName", "")
+                        if namespace == "rootio:language:python":
+                            for fixed_in in data["Vulnerability"].get("FixedIn", []):
+                                if fixed_in.get("Version") == "ROOTIO_UNAFFECTED":
```

Comment:
Please don't use magic / sentinel values like this. You want the vunnel to emit an actual version at which the vulnerability was fixed.